File: Algorithms/dpAlgorithms/LPsubstring.java
package dpAlgorithms;


public class LPsubstring 
{
    static void printSubStr(String str, int low, int high) {
        System.out.println(str.substring(low, high + 1));
    }
 
    static int longestPalSubstr(String str) {
        int n = str.length();
 
        boolean table[][] = new boolean[n][n];
 
        // All substrings of length 1 are palindromes
        int maxLength = 1;
        for (int i = 0; i < n; ++i)
            table[i][i] = true;
 
        // check for sub-string of length 2.
        int start = 0;
        for (int i = 0; i < n - 1; ++i) {
            if (str.charAt(i) == str.charAt(i + 1)) {
                table[i][i + 1] = true;
                start = i;
                maxLength = 2;
            }
        }
         
        // Check for lengths greater than 2. k is length
        // of substring
        for (int k = 3; k <= n; ++k) {
            for (int i = 0; i < n - k + 1; ++i)  {
                int j = i + k - 1;
                if (table[i + 1][j - 1] && str.charAt(i) == str.charAt(j)) {
                    table[i][j] = true;
                    if (k > maxLength) {
                        start = i;
                        maxLength = k;
                    }
                }
            }
        }
        System.out.print("Longest palindrome substring is: ");
        printSubStr(str, start, start+maxLength-1);
         
        return maxLength;
    }
    
    public static void main(String[] args) {
    	 
        String str = "jjkabcdsdcbajjk";
        System.out.println("Length is: " + longestPalSubstr(str));
    }
}

File: Algorithms/dpAlgorithms/LPsubstring.py
import sys

def printSubStr(st,low,high) :
    sys.stdout.write(st[low : high + 1])
    sys.stdout.flush()
    return ''
 

def longestPalSubstr(st) :
    n = len(st)
 
    table = [[0 for x in range(n)] for y in range(n)]
     
    # All substrings of length 1 are
    # palindromes
    maxLength = 1
    i = 0
    while (i < n) :
        table[i][i] = True
        i = i + 1
     
    # check for sub-string of length 2.
    start = 0
    i = 0
    while i < n - 1 :
        if (st[i] == st[i + 1]) :
            table[i][i + 1] = True
            start = i
            maxLength = 2
        i = i + 1
     
    # Check for lengths greater than 2. 
    # k is length of substring
    k = 3
    while k <= n :
        i = 0
        while i < (n - k + 1) :
            j = i + k - 1
			
            if (table[i + 1][j - 1] and
                      st[i] == st[j]) :
                table[i][j] = True
     
                if (k > maxLength) :
                    start = i
                    maxLength = k
            i = i + 1
        k = k + 1
    print "Longest palindrome substring is: ",printSubStr(st, start, start+maxLength-1)
 
    return maxLength
 
 
str = "jjkabcdsdcbajjk"
l = longestPalSubstr(str)
print "Length is: ", l